language/move-core/types/src/vm_status.rs

```diff
@@ -60,6 +60,31 @@ pub enum VMStatus {
 
     /// Indicates an `abort` from inside Move code. Contains the location of the abort and the code
     MoveAbort(AbortLocation, /* code */ u64),
+
+    /// Indicates an failure from inside Move code, where the VM could not continue exection, e.g.
+    /// dividing by zero or a missing resource
+    ExecutionFailure {
+        status_code: StatusCode,
+        location: AbortLocation,
+        function: u16,
+        code_offset: u16,
+    },
+}
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
+#[cfg_attr(any(test, feature = "fuzzing"), proptest(no_params))]
+pub enum RecordedStatus {
+    Executed,
+    OutOfGas,
+    MoveAbort(AbortLocation, /* code */ u64),
+    ExecutionFailure {
+        location: AbortLocation,
+        function: u16,
+        code_offset: u16,
+    },
+    VerificationError,
+    DeserializationError,
```

Comment:
@tzakian/@runtian-zhou: does this only happen when we fail to deserialize the script or its arguments? If so, might be slightly more intuitive to call it something like `MalformedScript` or `ScriptDeserializationError` to indicate that we parsed the "outer" transaction fine (hence why the tx was included/gas was charged), but not the script.

Reply:
that part is not well defined now and we should make sure it's properly handled.
That applies to verification errors as well. Both a deserialization and a verification errors on published modules would be disastrous, whereas in scripts or a module to publish perfectly expected.
I am not sure about the `DeserializationError` it seems too specific to me, but I am fine waiting the end of the PR and make a call later

Reply:
I think this should only happen when deserializing a user-submitted blob so I think `MalformedScript` wouldn't be bad, but @runtian-zhou  or @vgao1996  would probably know better than me. 

I also agree with Dario that any deserialization of on-chain data that fails is much more serious than a failure to deserialize user-submitted data  -- so should fall in the  invariant violation category.

Reply:
It's currently a StatusType that is marked as keep https://github.com/libra/libra/blob/15cc989a7ec611a07d860bf9a1e16add40c755a4/types/src/transaction/mod.rs#L582

These errors mostly come from that spot. Do we want new buckets? Are there things we are keeping we should discard?

Reply:
>  so should fall in the invariant violation category.

Invariant violations are discarded. Deserialization errors are not

Reply:
I think we are keeping things that we should discard, and we should adjust the buckets a bit. In particular:
* I think we should keep deserialization errors that arise when we deserialize any data that is submitted by the user (KEPT).
* I think any errors that arise when deserializing on-chain data (code or values) should be an invariant violation (DISCARDED).

Reply:
yes that is correct, and I can look into all of this once this is in if it helps

Reply:
> yes that is correct, and I can look into all of this once this is in if it helps

Sounds like a plan. Let's keep the behavior as is, and switch it after this lands. A ton going on right now. 
I can do it too, I'm just afraid of this PR being too big to handle

Reply:
Agreed. Also once this lands I'll do the plumbing of it in to the transactiontocommit and json-rpc side of things as well